src/python/pants/java/nailgun_protocol.py

```diff
@@ -232,7 +234,32 @@ def send_exit(cls, sock, payload=b''):
   @classmethod
   def send_pid(cls, sock, pid):
     """Send the PID chunk over the specified socket."""
-    cls.write_chunk(sock, ChunkType.PID, pid)
+    encoded_pid = cls.encode_int(pid)
+    cls.write_chunk(sock, ChunkType.PID, encoded_pid)
+
+  @classmethod
+  def encode_int(cls, obj):
+    """Verify the object is an int, and encode it to the right type of string for nailgun.
+
+    :param int obj: An integer to be encoded.
+    :raises: :class:`TypeError` if `obj` is not an integer.
+    :return: A representation of the int `obj` suitable to pass as the `payload` to send_exit().
+    """
+    if not isinstance(obj, int):
+      raise TypeError("cannot encode non-integer object in encode_int(): object was {} (type '{}')."
+                      .format(obj, type(obj)))
+    return str(obj).encode('ascii')
+
+  @classmethod
+  def send_exit_with_code(cls, sock, code):
+    """Send the Exit chunk over the specified socket, containing an integer return code."""
+    encoded_exit_status = cls.encode_int(code)
+    cls.send_exit(sock, payload=encoded_exit_status)
+
+  @classmethod
+  def encode_env_var_value(cls, obj):
+    """Encode the input so that it can be used as the value of an environment variable."""
+    return binary_type(str(obj))
```

Comment:
The method name is a bit misleading, I think... because this is going to go on the wire via nailgun, it's not immediately going to be used as an env var: instead, it must always output bytes. So maybe `encode_str_value`?

But regardless: this should probably be explicitly encoding as `utf-8`:
```
str(obj).encode('utf-8')
```

Reply:
This was indeed the error in the py3 shard -- I agree that it's not immediately going to be used as an env var, but I also want to make sure that everything that wants to be an environment variable's value goes through this function. Given that it is defined on the `NailgunProtocol` object, I think the intent could be made more clear by just editing the docstring to clarify that this is purely for the purposes of a nailgun env chunk (or just make the method to send those chunks call this one).

Reply:
I updated the docstring to clarify how the environment variables should be used.